home/models.py

```diff
@@ -1215,3 +1216,285 @@ class PageView(models.Model):
+    ):
+        previous_revision = self.get_latest_revision()
+
+        revision = super().save_revision(
+            user,
+            submitted_for_moderation,
+            approved_go_live_at,
+            changed,
+            log_action,
+            previous_revision,
+            clean,
+        )
+
+        try:
+            template_name = self.submit_whatsapp_template(previous_revision)
+        except Exception:
+            # Log the error to sentry and send error message to the user
+            logger.exception(f"Failed to submit template name:  {self.name}")
+            raise ValidationError("Failed to submit template")
+
+        if template_name:
+            revision.content["name"] = template_name
+            revision.save(update_fields=["content"])
+        return revision
+
+    def clean(self):
+        result = super().clean()
+        errors = {}
+
+        # The WA title is needed for all templates to generate a name for the template
+        if not self.name:
+            errors.setdefault("name", []).append(
+                ValidationError("All WhatsApp templates need a title.")
+            )
+
+        message = self.message
+
+        right_mismatch = re.findall(r"(?<!\{){[^{}]*}\}", message)
+        left_mismatch = re.findall(r"\{{[^{}]*}(?!\})", message)
+        mismatches = right_mismatch + left_mismatch
+
+        if mismatches:
+            errors.setdefault("message", []).append(
+                ValidationError(
+                    f"Please provide variables with matching braces. You provided {mismatches}."
+                )
+            )
+
+        vars_in_msg = re.findall(r"{{(.*?)}}", message)
+        non_digit_variables = [var for var in vars_in_msg if not var.isdecimal()]
+
+        if non_digit_variables:
+            errors.setdefault("message", []).append(
+                ValidationError(
+                    f"Please provide numeric variables only. You provided {non_digit_variables}."
+                )
+            )
+
+        # Check variable order
+        actual_digit_variables = [var for var in vars_in_msg if var.isdecimal()]
+        expected_variables = [str(j + 1) for j in range(len(actual_digit_variables))]
+        if actual_digit_variables != expected_variables:
+            errors.setdefault("message", []).append(
+                StreamBlockValidationError(
+                    {
+                        0: StreamBlockValidationError(
+                            {
+                                "message": ValidationError(
+                                    f'Variables must be sequential, starting with "{{1}}". Your first variable was "{actual_digit_variables}"'
+                                )
+                            }
+                        )
+                    }
+                )
+            )
+
+        if errors:
+            raise ValidationError(errors)
+
+        return result
+
+    @property
+    def whatsapp_template_fields(self):
```

Comment:
This appears to serve that same purpose as the `fields` property above. Do we need both? If so, what's the difference?

Reply:
Nah, I've remove `fields` this replaces it